arch/risc-v/src/mpfs/mpfs_ihc.c

```diff
@@ -0,0 +1,1276 @@
+
+/****************************************************************************
+ * Name: mpfs_ihc_init
+ *
+ * Description:
+ *   This initializes the Inter-Hart Communication (IHC) module.  Rptun is
+ *   used to simplify the integration of rpmsg and virtio.  This function
+ *   installs the proper interrupt handlers, installs a thread, and performs
+ *   all the required initialization tasks.
+ *
+ * Input Parameters:
+ *   None
+ *
+ * Returned Value:
+ *   OK on success, a nagated errno on error
+ *
+ ****************************************************************************/
+
+int mpfs_ihc_init(void)
+{
+  uint64_t  mhartid = riscv_mhartid();
+  FAR char *argv[3];
+  char      arg1[19];
+  uint32_t  rhartid;
+  int       ret;
+
+  /* We're IHC channel side B (slave) */
+
+#ifdef IHC_CHANNEL_SIDE_A
+  rhartid = CONTEXTB_HARTID;
+#else
+  rhartid = CONTEXTA_HARTID;
+#endif
+
+  /* Initialize IHC FPGA module registers to a known state */
+
+  mpfs_ihc_local_context_init((uint32_t)mhartid);
+  mpfs_ihc_local_remote_config((uint32_t)mhartid, rhartid);
+
+  /* Attach and enable the applicable irq */
+
+  ret = irq_attach(MPFS_IRQ_FABRIC_F2H_59, mpfs_ihc_interrupt, NULL);
+  if (ret == OK)
+    {
+      up_enable_irq(MPFS_IRQ_FABRIC_F2H_59);
+    }
+  else
+    {
+      ihcerr("ERROR: Not able to attach irq\n");
+      return ret;
+    }
+
+  /* Initialize and wait for the master. This will block until. */
+
+  ihcinfo("Waiting for the master online...\n");
+  ret = mpfs_rptun_init(MPFS_RPTUN_SHMEM_NAME, MPFS_RPTUN_CPU_NAME);
+  ihcinfo("..master is online\n");
+
+  /* Register callback to notify when rpmsg device is ready */
+
+  ret = rpmsg_register_callback(NULL, mpfs_rpmsg_device_created, NULL, NULL);
+  if (ret < 0)
+    {
+      ihcerr("ERROR: Not able to register rpmsg callback\n");
+      goto init_error;
+    }
+
+  /* Thread initialization */
+
+  snprintf(arg1, sizeof(arg1), "0x%" PRIxPTR,
+          (uintptr_t)g_mpfs_virtqueue_table);
+  argv[0] = (void *)"mpfs_ihc_thread";
+  argv[1] = arg1;
+  argv[2] = NULL;
+
+  ret = kthread_create("mpfs-rptun", CONFIG_RPTUN_PRIORITY,
+                       CONFIG_RPTUN_STACKSIZE, mpfs_rptun_thread, argv);
+  if (ret < 0)
+    {
+      ihcerr("ERROR: Not able to create a thread!\n");
+      rpmsg_unregister_callback(NULL, mpfs_rpmsg_device_created, NULL, NULL);
+      goto init_error;
+    }
+
+  return OK;
+
+init_error:
+
+  up_disable_irq(MPFS_IRQ_FABRIC_F2H_59);
+  return ret;
+}
+
+/****************************************************************************
+ * Name: up_addrenv_va_to_pa
+ *
+ * Description:
+ *   This is needed by openamp/libmetal/lib/system/nuttx/io.c:78. The
+ *   physical memory is mapped as virtual.
+ *
+ * Input Parameters:
+ *   va_
+ *
+ * Returned Value:
+ *   va
+ *
+ ****************************************************************************/
+
+uintptr_t up_addrenv_va_to_pa(FAR void *va_)
+{
+  uintptr_t va = (uintptr_t)va_;
```

Comment:
based on API description I think better to make
```
uintptr_t up_addrenv_va_to_pa(FAR void *va)
{
  uintptr_t pa = (uintptr_t)va;

  return pa;
}
```
or even
```
uintptr_t up_addrenv_va_to_pa(FAR void *va)
{
  return (uintptr_t)va;
}
```

Reply:
Thanks @pkarashchenko , looks like the PR was merged, so I'll address the review issues on a separate patch.